gateway-service/src/main/java/org/zowe/apiml/gateway/ws/WebSocketProxyServerHandler.java

```diff
@@ -205,7 +206,12 @@ public void handleMessage(WebSocketSession webSocketSession, WebSocketMessage<?>
         log.debug("handleMessage(session={},message={})", webSocketSession, webSocketMessage);
         WebSocketRoutedSession session = getRoutedSession(webSocketSession);
         if (session != null) {
-            session.sendMessageToServer(webSocketMessage);
+            try {
+                session.sendMessageToServer(webSocketMessage);
+            } catch (WebSocketException e) {
+                log.debug("Error sending WebSocket message. Closing session due to exception:", e);
```

Comment:
So should the `session.close()` be called here in order to close it properly? Or is the `afterConnectionClosed()` method called automatically?

Reply:
My understanding is that it will be called automatically

Reply:
Well, I'd add the `afterConnectionClosed()` call explicitly there instead of `routedSessions.remove(webSocketSession.getId());` because what it does is that it should properly close both WebSocketSession and associated WebSocketRoutedSession and then remove it from the `routedSessions` map. That seems to me cleaner than just removing from the map.

Reply:
I guess @weinfurt is right. The handler removes this context immediately after closing, see:

https://github.com/zowe/api-layer/blob/b5e4056b974b6d69fe0179028c523e0518c2e964/gateway-service/src/main/java/org/zowe/apiml/gateway/ws/WebSocketProxyServerHandler.java#L186-L200

I guess the soon socket will close the soon resource will release.

Reply:
Ok, yes I think we can explicitly call the `afterConnectionClosed` then. I will apply the change

Reply:
I've pushed the commit, please let me know how it looks like. I've used `SESSION_NOT_RELIABLE` as close status since I think it fits better.

Reply:
Maybe I am wrong, but I assume the error in sending a message could be from any reason. It is not meaning explicitly the session is closed. I guess it is waiting for GC that can then close the websocket. This change just supplements an automatic process. Why don't just close it and then the websocket implementation itself will call the method `afterConnectionClosed`?

```
    @Override
    public void handleMessage(WebSocketSession webSocketSession, WebSocketMessage<?> webSocketMessage)
        throws Exception {
        log.debug("handleMessage(session={},message={})", webSocketSession, webSocketMessage);
        WebSocketRoutedSession session = getRoutedSession(webSocketSession);
        if (session != null) {
            session.sendMessageToServer(webSocketMessage);
            try {
                session.close(CloseStatus.SESSION_NOT_RELIABLE);
            } catch (IOException e) {
                log.debug("cannot close WebSocket session (id = {})", webSocketSession.getId());
                routedSessions.remove(webSocketSession.getId());
            }
        }
    }
```

Btw, `afterConnectionClosed` itself solves the exception improperly (see removing not in close, but in finally):

```
    @Override
    public void afterConnectionClosed(WebSocketSession session, CloseStatus status) {
        log.debug("afterConnectionClosed(session={},status={})", session, status);
        try {
            session.close(status);

            WebSocketRoutedSession webSocketRoutedSession = getRoutedSession(session);
            if (webSocketRoutedSession != null) {
                webSocketRoutedSession.close(status);
            }
        } catch (NullPointerException | IOException e) {
            log.debug("Error closing WebSocket connection: {}", e.getMessage(), e);
        } finally {
            routedSessions.remove(session.getId());
        }
    }
```

Reply:
Generally, it should be improved more, because an exception on `session.close` could skip `webSocketRoutedSession.close`...

Reply:
Yes I agree, this could be also a potential source of memory leak. I agree with the suggestion to move the `routedSessions.remove()` call to finally clause.

Reply:
Thanks, this should work fine.